### ISSUE TITLE
CS9236 reported even when lambda parameters/return types are explicitly specified

### ISSUE THREAD

**Issue by roji** [dotnet/roslyn#79650]
Following up on https://github.com/dotnet/efcore/pull/36454, see below for a minimal repro in which CS9236 (Compiling requires binding the lambda expression at least 200 times) is reported). Note that all lambdas have their parameter and return types explicitly determined, but CS9236 is still reported.

This reproed for me with dotnet SDK 10.0.100-preview.6.25358.103.
 
```c#
void Produces_a_diagnostic()
{
    AssertQuery(
        IQueryable<IQueryable<IQueryable<IQueryable<Customer>>>> (ISetSource ss) => ss.Set<Customer>()
            .Select(IQueryable<IQueryable<IQueryable<Customer>>> (Customer outer) => ss.Set<Customer>()
                .Select(IQueryable<IQueryable<Customer>> (Customer c) => ss.Set<Customer>()
                                                                                            // CS9236 just below
                    .Select(IQueryable<Customer> (Customer cc) => ss.Set<Customer>().OrderBy(string (Customer inner) => inner.CustomerID)))));

    void AssertQuery<TResult>(Func<ISetSource, IQueryable<TResult>> query)
        => throw new NotImplementedException();
}

// Interestingly, if I switch to a regular array with IEnumerable instead of IQueryable,
// the diagnostic goes away (possibly because of the removal of the extra Set() call?):
void Does_not_produce_a_diagnostic()
{
    AssertQuery(
        IEnumerable<IEnumerable<IEnumerable<IEnumerable<Customer>>>> (Customer[] ss) => ss
            .Select(IEnumerable<IEnumerable<IEnumerable<Customer>>> (Customer outer) => ss
                .Select(IEnumerable<IEnumerable<Customer>> (Customer c) => ss
                    .Select(IEnumerable<Customer> (Customer cc) => ss.OrderBy(string (Customer inner) => inner.CustomerID)))));

    void AssertQuery<TResult>(Func<Customer[], IEnumerable<TResult>> query)
        => throw new NotImplementedException();
}

public interface ISetSource
{
    IQueryable<TEntity> Set<TEntity>()
        where TEntity : class;
}

public class Customer
{
    public string CustomerID { get; set; }
}
```

/cc @jaredpar @CyrusNajmabadi

**Comment by jaredpar** [comment]
Wonder if the issue is that we didn't add the same short circuits for expressions lambdas as normal lambdas. I can still reproduce the diagnostic when I fully specific every generic argument in that method.

**Comment by roji** [comment]
Yeah, could be related to the expression lambdas - but note that in https://github.com/dotnet/efcore/pull/36454 I did successfully get rid of most warnings (in expression lambdas) by explicitly typing the lambdas; the warnings remained in only two tests... So it might be a combination of factors...
